anta/tests/stp.py

```diff
@@ -0,0 +1,168 @@
+
+        for index, command in enumerate(self.instance_commands):
+            vlan_id = cast(Dict[str, str], command.template_params).get("vlan")
+            command_output = cast(Dict[str, Dict[Any, Any]], self.instance_commands[index].output)
+            logger.debug(f"dataset for vlan {vlan_id} is: {command_output}")
+
+            if command_output["spanningTreeVlanInstances"][vlan_id]["spanningTreeVlanInstance"]["protocol"] != mode:
+                self.result.is_failure(f"Wrong STP mode configured for VLAN {vlan_id}")
+
+
+class VerifySTPBlockedPorts(AntaTest):
+    """
+    Verifies there is no STP blocked ports.
+
+    Expected Results:
+        * success: The test will pass if there are NO ports blocked by STP.
+        * failure: The test will fail if there are ports blocked by STP.
+    """
+
+    name = "VerifySTPBlockedPorts"
+    description = "Verifies there is no STP blocked ports."
+    categories = ["stp"]
+    commands = [AntaTestCommand(command="show spanning-tree blockedports")]
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """
+        Run VerifySTPBlockedPorts validation
+        """
+        logger.debug(f"self.instance_commands is: {self.instance_commands}")
+
+        command_output = cast(Dict[str, Dict[Any, Any]], self.instance_commands[0].output)
+
+        if not (stp_instances := command_output["spanningTreeInstances"]):
+            self.result.is_success()
+        else:
+            for key, value in stp_instances.items():
+                stp_instances[key] = value.pop("spanningTreeBlockedPorts")
+            self.result.is_failure(f"The following ports are blocked by STP: {stp_instances}")
+
+
+class VerifySTPCounters(AntaTest):
+    """
+    Verifies there is no errors in STP BPDU packets.
+
+    Expected Results:
+        * success: The test will pass if there are NO STP BPDU packet errors under all interfaces participating in STP.
+        * failure: The test will fail if there are STP BPDU packet errors on one or more interface(s).
+    """
+
+    name = "VerifySTPCounters"
+    description = "Verifies there is no errors in STP BPDU packets."
+    categories = ["stp"]
+    commands = [AntaTestCommand(command="show spanning-tree counters")]
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Run VerifySTPBlockedPorts validation"""
+        logger.debug(f"self.instance_commands is: {self.instance_commands}")
+
+        command_output = cast(Dict[str, Dict[Any, Any]], self.instance_commands[0].output)
+
+        interfaces_with_errors = [
+            interface for interface, counters in command_output["interfaces"].items() if counters["bpduTaggedError"] or counters["bpduOtherError"] != 0
+        ]
+
+        if interfaces_with_errors:
+            self.result.is_failure(f"The following interfaces have STP BPDU packet errors: {interfaces_with_errors}")
+            return
+
+        self.result.is_success()
+
+
+class VerifySTPForwardingPorts(AntaTest):
```

Comment:
This does not work on my setup:
You can reach line 167 without having the `interfaces_not_forwarding` variable declared.

```
           DEBUG    Test VerifySTPForwardingPorts for device leaf2b: running test                                 models.py:226
           DEBUG    self.instance_commands is: [AntaTestCommand(command='show spanning-tree topology vlan 100        stp.py:154
                    status', version='latest', ofmt='json', output={'topologies': {}, 'unmappedVlans': []},
                    template=AntaTestTemplate(template='show spanning-tree topology vlan {vlan} status',
                    version='latest', ofmt='json'), template_params={'vlan': '100'})]
           DEBUG    dataset for vlan 100 is: {'topologies': {}, 'unmappedVlans': []}                                 stp.py:161
           ERROR    Exception raised for test VerifySTPForwardingPorts (on device leaf2b) - UnboundLocalError     models.py:229
                    (local variable 'interfaces_not_forwarding' referenced before assignment)
           DEBUG    Traceback (most recent call last):                                                            models.py:230
                      File "/mnt/lab/projects/anta/anta/models.py", line 227, in wrapper
                        function(self, **kwargs)
                      File "/mnt/lab/projects/anta/anta/tests/stp.py", line 167, in test
                        if interfaces_not_forwarding:
                    UnboundLocalError: local variable 'interfaces_not_forwarding' referenced before assignment
```